### PR TITLE
Implement the TransICD model

### PR DESCRIPTION
**Pull request by abheesht17**
Paper: https://arxiv.org/abs/2104.10652
Repository: https://github.com/biplob1ly/TransICD

### CODE REVIEW

**Comment by abheesht17** [comment]
**Transformer + Attention**

MIMIC-III 50 (old)

```
2022-05-17 20:44:35,012 — src.models.transicd.model — INFO — Initialising TransICD
2022-05-17 20:44:35,194 — src.trainers.base_trainer — INFO — Initializing BaseTrainer
2022-05-17 20:44:38,622 — src.trainers.base_trainer — INFO — Loaded checkpoint from best-11.pth
2022-05-17 20:44:38,628 — src.trainers.base_trainer — INFO — Use GPU
2022-05-17 20:44:38,628 — src.trainers.base_trainer — INFO — Evaluating on test dataset
2022-05-17 20:44:48,591 — src.trainers.base_trainer — INFO —    prec_at_5: 0.599769
2022-05-17 20:44:48,592 — src.trainers.base_trainer — INFO —    prec_at_8: 0.494072
2022-05-17 20:44:48,592 — src.trainers.base_trainer — INFO —     macro_f1: 0.505647
2022-05-17 20:44:48,592 — src.trainers.base_trainer — INFO —     micro_f1: 0.610448
2022-05-17 20:44:48,592 — src.trainers.base_trainer — INFO —    macro_auc: 0.887140
2022-05-17 20:44:48,592 — src.trainers.base_trainer — INFO —    micro_auc: 0.916317
2022-05-17 20:44:48,592 — src.trainers.base_trainer — INFO — Saving result on results/transicd_mimic3_50_old/test_result.json
```

MIMIC-III 50 (new)

```
2022-05-17 20:45:36,252 — src.trainers.base_trainer — INFO — Evaluating on test dataset
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO —    prec_at_5: 0.630365
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO —    prec_at_8: 0.514340
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO —     macro_f1: 0.581063
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO —     micro_f1: 0.663430
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO —    macro_auc: 0.910518
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO —    micro_auc: 0.933340
2022-05-17 20:45:53,393 — src.trainers.base_trainer — INFO — Saving result on results/transicd_mimic3_50/test_result.json
```

Will upload LDAM results soon.

File: src/models/transicd/model.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+            torch.arange(0, d_model, 2) * -(math.log(10000.0) / d_model)
+        )
+        positional_emb[:, 0::2] = torch.sin(position * div_term)
+        positional_emb[:, 1::2] = torch.cos(position * div_term)
+
+        # Revisit this step. In the official TransICD code, they take the
+        # transpose: positional_emb.unsqueeze(0).transpose(0, 1). Not sure if
+        # that makes sense. The shape changes from (bsz, max_len, d_model) to
+        # (max_len, bsz, d_model).
+        positional_emb = positional_emb.unsqueeze(0)
```
Comment (dalgu90) [inline src/models/transicd/model.py:198]: I've checked how `PositionalEncoding` works in the TransICD code, and it does work strangely.  
(Maybe we have to imitate this as well?)

**Review by dalgu90 — COMMENTED** [review]
Great work! I was surprised by the amount of work needed for the implementation of TransICD. Here are some comments and consideration

**Review by abheesht17 — COMMENTED** [review]
Thanks a lot for the review and the code loading analysis, @dalgu90!

**Review by dalgu90 — APPROVED** [review]

**Comment by abheesht17** [comment]
Thanks, @dalgu90, for the approval!

**Pull request merged**